Reaper split

sychan opened the pull request:
Change how the reaper is run so that there is an endpoint for reaping polled by the health checker. This can run in process or as a separate process.

sychan committed:
Reaper no longer gets scheduled, runs as main loop with --reaper flag

sychan committed:
Convert to reaping from endpoint call.

kkellerlbl commented on app.py:
```diff
@@ -10,12 +10,13 @@
 from datetime import datetime
 import json
 import hashlib
+import signal
```
Is signal used anywhere?  I didn't see it anywhere else in the PR.

kkellerlbl commented on app.py:
```diff
@@ -10,12 +10,13 @@
+import signal
 import manage_docker
 import manage_rancher
 from apscheduler.schedulers.background import BackgroundScheduler
```
Could we remove importing apscheduler (and code later that sets up the scheduler)?

kkellerlbl commented on app.py:
```diff
@@ -602,6 +600,21 @@ def narrative_status():
+    logger.info({"message": "Reaper endpoint called"})
+    try:
+        reaper()
+        resp = flask.Response("Reaper success")
```
Could we have the endpoint return the number of containers reaped, since reaper() now returns that info?

kkellerlbl commented on app.py:
```diff
@@ -602,6 +600,21 @@ def narrative_status():
     return(flask.Response(json.dumps(resp_doc), 200, mimetype='application/json'))
 
 
+@app.route("/reaper/", methods=['GET'])
```
One way we could secure this endpoint is with the admin role.  Then whatever requested the /reaper/ URL would need an appropriate KBase token.  I am honestly not sure what mode is better, using a token or hiding the container running the reaper.

kkellerlbl replied:
...actually if we're having Rancher call /reaper.../ as a health check it may not be able to provide a token.  I don't see any options for setting a header in the UI.

Alternatively we could have a check_mk HTTP rule to call /reaper..../.  check_mk does support setting headers.

kkellerlbl reviewed (commented):
Looks okay so far.  I made some minor comments but nothing blocking.

sychan committed:
Remove unused imports

kkellerlbl committed:
Update app.py

sychan committed:
Fix overzealous and underzealous deletions

sychan committed:
Respond with the number of containers reaped

sychan committed:
Merge branch 'reaper_split' into reaper_split_bugfix

sychan committed:
Merge pull request #35 from kbase/reaper_split_bugfix

reaper split bugfixes

kkellerlbl committed:
Add basic access control to /reaper/

Compare remote_addr to reaper_allowed_ip, forbid request if it doesn't match.  (Can use 0.0.0.0 to allow /reaper/ from anywhere.)

kkellerlbl committed:
Update app.py

Fix dumb typo.

kkellerlbl committed:
Update app.py

Use ipaddress module to allow specifying a subnet.

kkellerlbl committed:
Remove redundant line

kkellerlbl committed:
Merge pull request #36 from kbase/reaper_allowed_ip

Add basic access control to /reaper/
